File: backend/app/models/submissions.py
from app.extensions import db
from app.models.article import AI_INVOLVEMENT_VALUES

# A public story submission is reviewed by editorial staff before it can
# ever inform published content — it is never itself an Article. See
# STORY SUBMISSION VS ARTICLE in the task spec.
#
# submitted: received, not yet reviewed.
# reviewing: editorial team is assessing it.
# needs_information: additional details/materials requested from the submitter.
# shortlisted: potential fit, further review needed.
# approved: approved for editorial development (not yet an Article).
# converted: an Article draft has been created from this submission.
# published: the linked Article has actually been published (see
#   `is_published`/status-guard below — never set manually to imply
#   publication that hasn't happened).
# declined: not proceeding.
# withdrawn: submitter requested withdrawal, or the process ended that way.
# archived: historical record retained, out of active review.
SUBMISSION_STATUSES = (
    "submitted",
    "reviewing",
    "needs_information",
    "shortlisted",
    "approved",
    "converted",
    "published",
    "declined",
    "withdrawn",
    "archived",
)
_SUBMISSION_STATUS_CHECK_SQL = "status IN (" + ", ".join(f"'{s}'" for s in SUBMISSION_STATUSES) + ")"

# A controlled classification of what kind of story this is — editorial
# reuses/aligns with the existing Topics/Series taxonomy for subject
# matter, but this captures the narrative *shape* a submitter is offering,
# which topics/series alone don't express.
STORY_TYPES = (
    "personal_story",
    "career_journey",
    "leadership_story",
    "founder_story",
    "business_story",
    "community_impact",
    "starting_again",
    "overcoming_barriers",
    "women_doing_incredible_things",
    "women_leading_organizations",
    "workplace_story",
    "opportunity_achievement",
    "other",
)
_STORY_TYPE_CHECK_SQL = "story_type IS NULL OR story_type IN (" + ", ".join(
    f"'{t}'" for t in STORY_TYPES
) + ")"

# Where the submitted text actually came from — informs editorial
# handling, never automated copyright verification.
CONTENT_ORIGINS = ("first_person", "on_behalf_of", "previously_published", "adapted")
_CONTENT_ORIGIN_CHECK_SQL = "content_origin IS NULL OR content_origin IN (" + ", ".join(
    f"'{o}'" for o in CONTENT_ORIGINS
) + ")"

_AI_INVOLVEMENT_CHECK_SQL = "ai_involvement IN (" + ", ".join(f"'{v}'" for v in AI_INVOLVEMENT_VALUES) + ")"

# Simple internal indicators, not a fact-checking platform.
VERIFICATION_STATUSES = ("unverified", "verification_needed", "verified")
_VERIFICATION_STATUS_CHECK_SQL = "verification_status IN (" + ", ".join(
    f"'{v}'" for v in VERIFICATION_STATUSES
) + ")"

# When a story is about someone other than the submitter, staff need a
# clear indicator of whether the subject's own consent is confirmed
# before anything about her can be published — never inferred, always
# staff-recorded.
SUBJECT_PERMISSION_STATUSES = ("not_applicable", "unknown", "needs_confirmation", "confirmed")
_SUBJECT_PERMISSION_CHECK_SQL = "subject_permission_status IN (" + ", ".join(
    f"'{s}'" for s in SUBJECT_PERMISSION_STATUSES
) + ")"

submission_topics = db.Table(
    "submission_topics",
    db.Column("submission_id", db.Integer, db.ForeignKey("story_submissions.id", ondelete="CASCADE"), primary_key=True),
    db.Column("topic_id", db.Integer, db.ForeignKey("topics.id", ondelete="CASCADE"), primary_key=True),
)


class StorySubmission(db.Model):
    """A story received from the public — reviewed and classified by
    editorial staff, and only ever turned into an Article through a
    deliberate "Create Article Draft" handoff (never auto-published). One
    evolving record from first submission through to Converted/Published/
    Declined/Withdrawn/Archived, matching the PartnershipInquiry/Member
    pattern used elsewhere rather than separate lead/converted tables.
    """

    __tablename__ = "story_submissions"
    __table_args__ = (
        db.CheckConstraint(_SUBMISSION_STATUS_CHECK_SQL, name="ck_story_submissions_status"),
        db.CheckConstraint(_STORY_TYPE_CHECK_SQL, name="ck_story_submissions_story_type"),
        db.CheckConstraint(_CONTENT_ORIGIN_CHECK_SQL, name="ck_story_submissions_content_origin"),
        db.CheckConstraint(_AI_INVOLVEMENT_CHECK_SQL, name="ck_story_submissions_ai_involvement"),
        db.CheckConstraint(_VERIFICATION_STATUS_CHECK_SQL, name="ck_story_submissions_verification_status"),
        db.CheckConstraint(_SUBJECT_PERMISSION_CHECK_SQL, name="ck_story_submissions_subject_permission"),
    )

    id = db.Column(db.Integer, primary_key=True)
    # Human-friendly, unique, immutable — e.g. "WSF-STORY-2026-00125".
    # Generated once via app/services/submissions.py:generate_submission_reference
    # right after the row has an id, same pattern as Order.reference.
    reference = db.Column(db.String(30), unique=True, nullable=True, index=True)

    # --- Submitter (About You) ---
    first_name = db.Column(db.String(120), nullable=False)
    last_name = db.Column(db.String(120), nullable=False)
    email = db.Column(db.String(255), nullable=False, index=True)
    country_code = db.Column(db.String(10), db.ForeignKey("countries.code"), nullable=True)
    city = db.Column(db.String(120))
    professional_title = db.Column(db.String(200))
    organization_name = db.Column(db.String(200))
    linkedin_url = db.Column(db.String(500))
    website_url = db.Column(db.String(500))

    # --- Story ---
    title = db.Column(db.String(300), nullable=False)
    summary = db.Column(db.Text)
    body = db.Column(db.Text, nullable=False)  # the original submitted material — never overwritten
    why_it_matters = db.Column(db.Text)
    key_lessons = db.Column(db.Text)
    story_type = db.Column(db.String(40))

    # --- Subject ---
    subject_is_submitter = db.Column(db.Boolean, nullable=False, default=True)
    subject_name = db.Column(db.String(200))
    subject_relationship = db.Column(db.String(200))
    subject_permission_status = db.Column(db.String(20), nullable=False, default="not_applicable")

    # --- Provenance / content origin ---
    content_origin = db.Column(db.String(20))
    previous_publication_url = db.Column(db.String(500))
    ai_involvement = db.Column(db.String(30), nullable=False, default="none")
    ai_provenance_note = db.Column(db.Text)

    # --- Consent / rights (never exposed publicly) ---
    consent_review_given = db.Column(db.Boolean, nullable=False, default=False)
    consent_contact_given = db.Column(db.Boolean, nullable=False, default=False)
    consent_accuracy_confirmed = db.Column(db.Boolean, nullable=False, default=False)
    consent_media_rights_confirmed = db.Column(db.Boolean, nullable=True)
    consent_recorded_at = db.Column(db.DateTime(timezone=True))

    # Separate, optional, unchecked-by-default — see NEWSLETTER CONSENT.
    # The live subscription lives in NewsletterSubscriber; this is only an
    # audit record of what the submitter asked for at submission time.
    newsletter_opt_in = db.Column(db.Boolean, nullable=False, default=False)

    # --- Editorial classification (staff-set; never required of the public submitter) ---
    series_id = db.Column(db.Integer, db.ForeignKey("series.id"), nullable=True)
    recommended_format = db.Column(db.String(120))
    verification_status = db.Column(db.String(20), nullable=False, default="unverified")
    permission_followup_required = db.Column(db.Boolean, nullable=False, default=False)
    media_followup_required = db.Column(db.Boolean, nullable=False, default=False)
    information_requested_note = db.Column(db.Text)  # what's needed, when status = needs_information

    # --- Review / status ---
    status = db.Column(db.String(20), nullable=False, default="submitted")
    editorial_assessment = db.Column(db.Text)
    assigned_editor_id = db.Column(db.Integer, db.ForeignKey("users.id"), nullable=True)
    reviewed_at = db.Column(db.DateTime(timezone=True))

    # --- Linked records (all optional, all editor-controlled — never auto-created) ---
    person_id = db.Column(db.Integer, db.ForeignKey("people.id"), nullable=True)
    organization_id = db.Column(db.Integer, db.ForeignKey("organizations.id"), nullable=True)

    acquisition = db.Column(db.JSON)
    submitted_at = db.Column(db.DateTime(timezone=True), server_default=db.func.now(), nullable=False)
    updated_at = db.Column(
        db.DateTime(timezone=True), server_default=db.func.now(), onupdate=db.func.now(), nullable=False
    )

    country = db.relationship("Country", foreign_keys=[country_code])
    series = db.relationship("Series", foreign_keys=[series_id])
    assigned_editor = db.relationship("User", foreign_keys=[assigned_editor_id])
    person = db.relationship("Person", foreign_keys=[person_id])
    organization = db.relationship("Organization", foreign_keys=[organization_id])
    topics = db.relationship("Topic", secondary=submission_topics)
    notes = db.relationship(
        "SubmissionNote", backref="submission", cascade="all, delete-orphan", order_by="SubmissionNote.created_at"
    )
    media_items = db.relationship(
        "SubmissionMedia", backref="submission", cascade="all, delete-orphan", order_by="SubmissionMedia.created_at"
    )
    # `resulting_article` is defined as a backref from Article.source_submission_id
    # (see app/models/article.py) — the authoritative FK lives on Article.

    @property
    def full_name(self):
        return f"{self.first_name} {self.last_name}".strip()

    def sync_published_state(self):
        """A submission may only read as Published when its linked Article
        actually is — never set manually to imply a publication that
        hasn't happened. Called after loading/whenever the linked
        Article's status might have changed.
        """
        if self.status == "published" and (not self.resulting_article or self.resulting_article.status != "published"):
            self.status = "converted"
        elif self.status == "converted" and self.resulting_article and self.resulting_article.status == "published":
            self.status = "published"


class SubmissionNote(db.Model):
    """An internal, staff-only note on a submission — never returned by any
    public response. Same append-only pattern as PartnershipNote/MemberNote.
    """

    __tablename__ = "submission_notes"

    id = db.Column(db.Integer, primary_key=True)
    submission_id = db.Column(db.Integer, db.ForeignKey("story_submissions.id", ondelete="CASCADE"), nullable=False)
    user_id = db.Column(db.Integer, db.ForeignKey("users.id"), nullable=False)
    body = db.Column(db.Text, nullable=False)
    created_at = db.Column(db.DateTime(timezone=True), server_default=db.func.now(), nullable=False)

    user = db.relationship("User", foreign_keys=[user_id])


class SubmissionMedia(db.Model):
    """One optional image attached to a submission, with the rights/credit
    context editorial needs before ever reusing it — reuses the existing
    Media model/upload service for storage (see the public
    `POST /submissions/media` route), never a second file-storage system.
    Kept as its own association row (not a bare many-to-many) because the
    caption/credit/rights confirmation are specific to this submission's
    use of the image, not the shared Media row's global metadata.
    """

    __tablename__ = "submission_media"

    id = db.Column(db.Integer, primary_key=True)
    submission_id = db.Column(db.Integer, db.ForeignKey("story_submissions.id", ondelete="CASCADE"), nullable=False)
    media_id = db.Column(db.Integer, db.ForeignKey("media.id"), nullable=False)
    caption = db.Column(db.String(500))
    credit = db.Column(db.String(255))
    rights_confirmed = db.Column(db.Boolean, nullable=False, default=False)
    created_at = db.Column(db.DateTime(timezone=True), server_default=db.func.now(), nullable=False)

    media = db.relationship("Media", foreign_keys=[media_id])

File: frontend/src/api/submissions.js
import { apiClient, USE_MOCK } from './client'
import { delay } from './mockUtils'
import { mapMediaRef } from '../utils/media'

// No dedicated Story Submissions mock dataset exists yet (this module is
// new) — a small self-contained fixture keeps mock mode functional
// without growing the shared mock/admin.js file for a shape nothing
// else reads.
const MOCK_SUBMISSIONS = []

export function mapSubmissionListItem(s) {
  if (!s) return null
  return {
    id: s.id,
    reference: s.reference || '',
    fullName: s.full_name || `${s.first_name || ''} ${s.last_name || ''}`.trim(),
    email: s.email || '',
    country: s.country ? { code: s.country.code, name: s.country.name } : null,
    title: s.title || '',
    storyType: s.story_type || '',
    status: s.status || 'submitted',
    assignedEditor: s.assigned_editor ? { id: s.assigned_editor.id, fullName: s.assigned_editor.full_name } : null,
    submittedAt: s.submitted_at || null,
    updatedAt: s.updated_at || null,
  }
}

export function mapSubmission(s) {
  if (!s) return null
  return {
    id: s.id,
    reference: s.reference || '',
    firstName: s.first_name || '',
    lastName: s.last_name || '',
    fullName: s.full_name || `${s.first_name || ''} ${s.last_name || ''}`.trim(),
    email: s.email || '',
    country: s.country ? { code: s.country.code, name: s.country.name } : null,
    countryCode: s.country?.code || '',
    city: s.city || '',
    professionalTitle: s.professional_title || '',
    organizationName: s.organization_name || '',
    linkedinUrl: s.linkedin_url || '',
    websiteUrl: s.website_url || '',

    title: s.title || '',
    summary: s.summary || '',
    body: s.body || '',
    whyItMatters: s.why_it_matters || '',
    keyLessons: s.key_lessons || '',
    storyType: s.story_type || '',
    topics: Array.isArray(s.topics) ? s.topics.map((t) => ({ id: t.id, slug: t.slug, name: t.name })) : [],
    topicSlugs: Array.isArray(s.topics) ? s.topics.map((t) => t.slug) : [],
    series: s.series ? { id: s.series.id, slug: s.series.slug, name: s.series.name } : null,
    seriesId: s.series_id || null,

    subjectIsSubmitter: s.subject_is_submitter ?? true,
    subjectName: s.subject_name || '',
    subjectRelationship: s.subject_relationship || '',
    subjectPermissionStatus: s.subject_permission_status || 'not_applicable',

    contentOrigin: s.content_origin || '',
    previousPublicationUrl: s.previous_publication_url || '',
    aiInvolvement: s.ai_involvement || 'none',
    aiProvenanceNote: s.ai_provenance_note || '',

    consentReviewGiven: !!s.consent_review_given,
    consentContactGiven: !!s.consent_contact_given,
    consentAccuracyConfirmed: !!s.consent_accuracy_confirmed,
    consentMediaRightsConfirmed: s.consent_media_rights_confirmed ?? null,
    consentRecordedAt: s.consent_recorded_at || null,
    newsletterOptIn: !!s.newsletter_opt_in,

    recommendedFormat: s.recommended_format || '',
    verificationStatus: s.verification_status || 'unverified',
    permissionFollowupRequired: !!s.permission_followup_required,
    mediaFollowupRequired: !!s.media_followup_required,
    informationRequestedNote: s.information_requested_note || '',

    status: s.status || 'submitted',
    editorialAssessment: s.editorial_assessment || '',
    assignedEditor: s.assigned_editor ? { id: s.assigned_editor.id, fullName: s.assigned_editor.full_name } : null,
    reviewedAt: s.reviewed_at || null,

    person: s.person ? { id: s.person.id, slug: s.person.slug, name: s.person.name } : null,
    personId: s.person_id || null,
    organization: s.organization ? { id: s.organization.id, slug: s.organization.slug, name: s.organization.name } : null,
    organizationId: s.organization_id || null,

    mediaItems: Array.isArray(s.media_items)
      ? s.media_items.map((m) => ({
          id: m.id,
          media: mapMediaRef(m.media),
          caption: m.caption || '',
          credit: m.credit || '',
          rightsConfirmed: !!m.rights_confirmed,
        }))
      : [],

    resultingArticle: s.resulting_article
      ? { id: s.resulting_article.id, slug: s.resulting_article.slug, title: s.resulting_article.title, status: s.resulting_article.status }
      : null,

    notes: Array.isArray(s.notes)
      ? s.notes.map((n) => ({ id: n.id, body: n.body, user: n.user?.full_name || null, createdAt: n.created_at }))
      : [],

    submittedAt: s.submitted_at || null,
    updatedAt: s.updated_at || null,
  }
}

// ---------------------------------------------------------------------------
// Public
// ---------------------------------------------------------------------------

function buildSubmissionPayload(payload) {
  return {
    firstName: payload.firstName,
    lastName: payload.lastName,
    email: payload.email,
    countryCode: payload.countryCode,
    city: payload.city || undefined,
    professionalTitle: payload.professionalTitle || undefined,
    organizationName: payload.organizationName || undefined,
    linkedinUrl: payload.linkedinUrl || undefined,
    websiteUrl: payload.websiteUrl || undefined,

    title: payload.title,
    summary: payload.summary,
    body: payload.body,
    whyItMatters: payload.whyItMatters || undefined,
    keyLessons: payload.keyLessons || undefined,
    storyType: payload.storyType || undefined,
    topicSlugs: payload.topicSlugs,

    subjectIsSubmitter: payload.subjectIsSubmitter,
    subjectName: payload.subjectIsSubmitter ? undefined : payload.subjectName || undefined,
    subjectRelationship: payload.subjectIsSubmitter ? undefined : payload.subjectRelationship || undefined,

    contentOrigin: payload.contentOrigin || undefined,
    previousPublicationUrl: payload.previousPublicationUrl || undefined,
    aiInvolvement: payload.aiInvolvement || 'none',
    aiProvenanceNote: payload.aiProvenanceNote || undefined,

    media: (payload.media || []).map((m) => ({
      mediaId: m.mediaId,
      caption: m.caption || undefined,
      credit: m.credit || undefined,
      rightsConfirmed: m.rightsConfirmed,
    })),

    consentReviewGiven: payload.consentReviewGiven,
    consentContactGiven: payload.consentContactGiven,
    consentAccuracyConfirmed: payload.consentAccuracyConfirmed,
    consentMediaRightsConfirmed: payload.media?.length ? payload.consentMediaRightsConfirmed : undefined,
    newsletterOptIn: payload.newsletterOptIn,

    acquisition: payload.acquisition,
  }
}

export async function submitStorySubmission(payload) {
  if (!USE_MOCK) {
    try {
      const { data } = await apiClient.post('/submissions', buildSubmissionPayload(payload))
      return {
        success: true,
        reference: data.reference,
        message: `Your story has been received for editorial review (reference ${data.reference}). We can't guarantee publication, but our editorial team reviews every submission.`,
      }
    } catch (err) {
      return { success: false, message: err.apiError?.message || 'Something went wrong. Please try again.' }
    }
  }
  return delay({
    success: true,
    reference: 'WSF-STORY-2026-00001',
    message: "Your story has been received for editorial review (reference WSF-STORY-2026-00001). We can't guarantee publication, but our editorial team reviews every submission.",
  }, 500)
}

// Optional image upload used solely to attach a photo to a story
// submission before it's sent — reuses the same validated upload flow as
// the admin media library, but through a public, permission-free
// endpoint (see backend SubmissionMediaUploadResource).
export async function uploadSubmissionMedia(file, metadata = {}) {
  if (!USE_MOCK) {
    const formData = new FormData()
    formData.append('file', file)
    if (metadata.caption) formData.append('caption', metadata.caption)
    if (metadata.credit) formData.append('credit', metadata.credit)
    const { data } = await apiClient.post('/submissions/media', formData)
    return { id: data.id, publicUrl: data.publicUrl }
  }
  const objectUrl = URL.createObjectURL(file)
  return delay({ id: `mock-${Date.now()}`, publicUrl: objectUrl }, 400)
}

// ---------------------------------------------------------------------------
// Admin
// ---------------------------------------------------------------------------

export async function fetchSubmissions(params = {}) {
  if (!USE_MOCK) {
    const { data } = await apiClient.get('/submissions', { params })
    return { ...data, items: data.items.map(mapSubmissionListItem) }
  }
  return delay({ items: MOCK_SUBMISSIONS, meta: { page: 1, per_page: 20, total: 0, total_pages: 0, has_next: false, has_prev: false } })
}

export async function fetchSubmission(id) {
  if (!USE_MOCK) {
    const { data } = await apiClient.get(`/submissions/${id}`)
    return mapSubmission(data)
  }
  return delay(null)
}

export async function updateSubmission(id, payload) {
  const body = {
    storyType: payload.storyType || undefined,
    topicSlugs: payload.topicSlugs,
    seriesId: payload.seriesId === '' || payload.seriesId == null ? null : Number(payload.seriesId),
    recommendedFormat: payload.recommendedFormat || undefined,
    verificationStatus: payload.verificationStatus || undefined,
    permissionFollowupRequired: payload.permissionFollowupRequired,
    mediaFollowupRequired: payload.mediaFollowupRequired,
    informationRequestedNote: payload.informationRequestedNote || undefined,
    editorialAssessment: payload.editorialAssessment || undefined,
    subjectPermissionStatus: payload.subjectPermissionStatus || undefined,
    personId: payload.personId === '' || payload.personId == null ? null : Number(payload.personId),
    organizationId: payload.organizationId === '' || payload.organizationId == null ? null : Number(payload.organizationId),
  }
  const { data } = await apiClient.patch(`/submissions/${id}`, body)
  return mapSubmission(data)
}

export async function updateSubmissionStatus(id, status) {
  const { data } = await apiClient.patch(`/submissions/${id}/status`, { status })
  return mapSubmission(data)
}

export async function assignSubmissionEditor(id, editorId) {
  const { data } = await apiClient.post(`/submissions/${id}/assign`, { editorId: editorId || null })
  return mapSubmission(data)
}

export async function addSubmissionNote(id, body) {
  const { data } = await apiClient.post(`/submissions/${id}/notes`, { body })
  return mapSubmission(data)
}

export async function deleteSubmission(id) {
  await apiClient.delete(`/submissions/${id}`)
}

export async function fetchSubmissionHistory(id) {
  if (!USE_MOCK) {
    const { data } = await apiClient.get(`/submissions/${id}/history`)
    return data
  }
  return delay([])
}

export async function convertSubmissionToArticle(id, { authorId, includeMedia = true }) {
  const { data } = await apiClient.post(`/submissions/${id}/convert-to-article`, {
    authorId: Number(authorId),
    includeMedia,
  })
  return { submission: mapSubmission(data.submission), article: data.article }
}
